src/extension/css/src/test/java/org/geoserver/rest/catalog/CssStyleControllerTest.java

```diff
@@ -0,0 +1,167 @@
+    }
+
+    @Test
+    public void testRawPutCSS() throws Exception {
+        // step 1 create style info with correct format
+        Catalog cat = getCatalog();
+        assertNull("foo not available", cat.getStyleByName("foo"));
+
+        String xml = "<style>" + "<name>foo</name>" + "<format>css</format>"
+                + "<filename>foo.css</filename>" + "</style>";
+        MockHttpServletResponse response = postAsServletResponse(
+                RestBaseController.ROOT_PATH + "/styles", xml);
+        assertEquals(201, response.getStatus());
+        assertNotNull(cat.getStyleByName("foo"));
+
+        String content = newCSS();
+        response = putAsServletResponse("/rest/styles/foo?raw=true", content, CssHandler.MIME_TYPE);
+        assertEquals(200, response.getStatus());
+
+        GeoServerResourceLoader resources = catalog.getResourceLoader();
+
+        Resource resource = resources.get("/styles/foo.css");
+
+        String definition = new String(resource.getContents());
+        assertTrue("is css", definition.contains("stroke: red"));
+
+        StyleInfo styleInfo = catalog.getStyleByName("foo");
+        Style s = styleInfo.getStyle();
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+
+        SLDHandler handler = new SLDHandler();
+        handler.encode(Styles.sld(s), SLDHandler.VERSION_10, false, out);
+        content = new String(out.toByteArray());
+        assertTrue(content.contains("<sld:Name>foo</sld:Name>"));
+        catalog.remove(styleInfo);
+    }
+
+    @Test
+    public void testPutCSS() throws Exception {
+        // step 1 create style info with correct format
+        Catalog cat = getCatalog();
+        assertNull(cat.getStyleByName("bar"));
+
+        String xml = "<style>" + "<name>bar</name>" + "<format>css</format>"
+                + "<filename>bar.css</filename>" + "</style>";
+
+        MockHttpServletResponse response = postAsServletResponse(
+                RestBaseController.ROOT_PATH + "/styles", xml);
+        assertEquals(201, response.getStatus());
+        assertNotNull(cat.getStyleByName("bar"));
+
+        String content = newCSS();
+        response = putAsServletResponse("/rest/styles/bar", content, CssHandler.MIME_TYPE);
+        assertEquals(200, response.getStatus());
+
+        GeoServerResourceLoader resources = catalog.getResourceLoader();
+
+        Resource resource = resources.get("/styles/bar.css");
+
+        String definition = new String(resource.getContents());
+        assertTrue("is css", definition.contains("stroke: red"));
+
+        StyleInfo styleInfo = catalog.getStyleByName("bar");
+        Style s = styleInfo.getStyle();
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+
+        SLDHandler handler = new SLDHandler();
+        handler.encode(Styles.sld(s), SLDHandler.VERSION_10, false, out);
+        content = new String(out.toByteArray());
+        assertTrue(content.contains("<sld:Name>bar</sld:Name>"));
+
```

Comment:
Thanks for confirming this work, there is a new feature where CSS "validation" is performed, it should catch and reject case where a CSS file is uploaded that cannot generate a correct Style.